src/utilities/MassIntegrator.cpp

```diff
@@ -0,0 +1,309 @@
+#include "tbox/Utilities.h"
+
+#include <array>
+#include <limits>
+#include <memory>
+#include <ostream>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "ibamr/namespaces.h" // IWYU pragma: keep
+
+namespace SAMRAI
+{
+namespace solv
+{
+template <int DIM>
+class RobinBcCoefStrategy;
+} // namespace solv
+} // namespace SAMRAI
+
+/////////////////////////////// NAMESPACE ////////////////////////////////////
+
+namespace IBAMR
+{
+/////////////////////////////// STATIC ///////////////////////////////////////
+
+namespace
+{
+// NOTE: The number of ghost cells required by the convection scheme depends
+// on the chosen convective limiter, which will be set via input file
+static const int GPPMG = 4;
+static const int GCUIG = 3;
+
+// Timers.
+static Timer* t_apply_convective_operator;
+static Timer* t_integrate;
+static Timer* t_initialize_integrator;
+static Timer* t_deallocate_integrator;
+} // namespace
+
+/////////////////////////////// PUBLIC ///////////////////////////////////////
+MassIntegrator::MassIntegrator(std::string object_name, Pointer<Database> input_db)
+    : d_object_name(std::move(object_name)), d_u_bc_coefs(NDIM), d_rho_bc_coefs(NDIM)
+{
+    if (input_db)
+    {
+        if (input_db->keyExists("bdry_extrap_type"))
+        {
+            d_density_bdry_extrap_type = input_db->getString("bdry_extrap_type");
+        }
+        if (input_db->keyExists("density_bdry_extrap_type"))
+        {
+            d_density_bdry_extrap_type = input_db->getString("density_bdry_extrap_type");
+        }
+        if (input_db->keyExists("convective_limiter"))
+        {
+            d_density_convective_limiter =
+                IBAMR::string_to_enum<LimiterType>(input_db->getString("convective_limiter"));
+        }
+        if (input_db->keyExists("density_convective_limiter"))
+        {
+            d_density_convective_limiter =
+                IBAMR::string_to_enum<LimiterType>(input_db->getString("density_convective_limiter"));
+        }
+
+        if (input_db->keyExists("density_time_stepping_type"))
+        {
+            d_density_time_stepping_type =
+                IBAMR::string_to_enum<TimeSteppingType>(input_db->getString("density_time_stepping_type"));
+        }
+        if (input_db->keyExists("enable_logging"))
+        {
+            d_enable_logging = input_db->getBool("enable_logging");
+        }
+    }
+
+    switch (d_density_convective_limiter)
+    {
+    case PPM:
+        d_density_limiter_gcw = GPPMG;
+        break;
+    case CUI:
+        d_density_limiter_gcw = GCUIG;
+        break;
+    default:
+        TBOX_ERROR(
+            "MassIntegrator::"
+            "MassIntegrator():\n"
+            << "  unsupported density convective limiter: "
+            << IBAMR::enum_to_string<LimiterType>(d_density_convective_limiter) << " \n"
+            << "  valid choices are: PPM, CUI\n");
+    }
+
+    switch (d_density_time_stepping_type)
+    {
+    case FORWARD_EULER:
```

Comment:
Should these time stepping details go into the subclasses? It seems like the intent of the design is to let different implementations provide different time stepping schemes.

Reply:
FWIW, this is also the reason why I commented on the SSP implementation being derived from the non-SSP implementation instead of the ABC. I wonder if it would be better to have an abstract RK base class that only provide forward Euler, and then have two different concrete implementations (one that adds on the explicit midpoint rule, the other that adds on the SSP RK schemes).

Reply:
One technicality is the velocity field used in the intermediate stages of RK and SSP-RK. While the RK is simple $0.5*(u^{n+1} + u^{n})$, SSP-RK employs a more complicated interpolation of $u^{n+1}, u^{n}, u^{n-1}$ to achieve full accuracy. Hence we just separated them out.

Reply:
The interpolation scheme for SSP-RK utilizing $u^{n-1}, u^{n}, u^{n+1}$ is explained in our JCP paper:
https://www.sciencedirect.com/science/article/pii/S0021999119302256

In particular see Eqs. (43)-(46) and 3rd order (a.k.a "full") accuracy in $\rho$ by retaining $u^{n-1}$ in Fig. 15.

Reply:
Hrmmm. I don't remember this part of the paper. (Obviously. 😀) If the time stepping for $u$ is globally second-order, I am not sure if there is any benefit to overall order-of-accuracy to performing a local third-order reconstruction.

(The way I would think about it is that you cannot take the second-order accurate values $u^{n-1}$, $u^{n}$, and $u^{n+1}$ and perform a quadratic interpolation to recover a third-order accurate approximation to $u$ at $t \in [t^{n-1}, t^{n+1}]$. The reconstruction is still just second-order accurate.)

What happens if you omit using $u^{n-1}$ here? Is it important to use SSP-RK3 vs. SSP-RK2? Since everything else is solved using 2nd-order time stepping, I don't think using higher-order for the density equation should impact the order of accuracy, although these choices may impact the error constant. Omitting $u^{n-1}$ could also improve stability, and I think that the stability properties of SSP-RK2 and the 3-stage SSP-RK3 schemes are the same. (On the other hand, I think that 4-stage SSP-RK3 is more stable than 2-stage SSP-RK2.)

Typically, the main reason to use multi-step is to _avoid_ doing multiple stages (i.e., cycles). The usual tradeoff is typically in terms of stability. If I am doing multi-step for any part of the problem, I would try to do it for the whole thing, so that I only need to do a single linear solve per time step.

Note that there are SSP multistep methods --- see, e.g., https://arxiv.org/abs/1504.04107 and references therein. Maybe we should be using these instead of AB2 in the constant-coefficient solvers... 😁

Reply:
I think we were targeting accuracy of density, velocity, and pressure altogether. This particular interpolation scheme gave the maximum accuracy (and stability for a very high density ratio case) with ssp-rk3. Perhaps for practical problems, it does not matter. These days we are just using RK2 for density and finding it equally robust. Anyways, it is there as a valid option to try.

Reply:
SSK-RK2 is simply the explicit trapezoidal rule --- if it doesn't make any difference if you use SSP-RK2 or SSP-RK3, then do we need both the RK and SSP-RK classes?

Reply:
Note also that omitting the lagged value of the velocity will save some memory, which you have mentioned concerns about here. 😁

Reply:
BTW, to be clear, my intuition here is that SSP-RK2 without multistep extrapolation will be more stable than SSP-RK3 with multistep extrapolation.

Reply:
Here we are concerned with a different type of instability, which is mass-momentum inconsistency. The integration of rho with any sort of reasonable time stepping scheme is fine. The main point is using the same mass flux in the convective operator of the momentum equation. In this regard, rk, ssp-rk2/3 are all reasonable choices, because we utilize the same mass flux in the two advective operators.

I would like to keep ssp-rk2/3 because there was this paper that debated one needs to the same integrator in momentum eq. as in mass equation. Our implementation clearly shows that this is not true!

Reply:
IIRC ssp-rk3 was more stable than ssp-rk2 on some extremely high density ratio cases. @nishantnangia may recall better. I think this was the reason we stuck with ssp-rk3 in the paper (and even afterwards).